vm/devices/user_driver/src/dma.rs

```diff
@@ -0,0 +1,31 @@
+
+#[derive(Debug)]
+pub enum DmaError {
+    InitializationFailed,
+    MapFailed,
+    UnmapFailed,
+    PinFailed,
+    BounceBufferFailed,
+}
+
+// Structure encapsulating the result of a DMA mapping operation
+pub struct DmaTransactionHandler {
+    pub transactions: Vec<DmaTransaction>,
+}
+
+// Structure representing a DMA transaction with address and metadata
+pub struct DmaTransaction {
+    pub original_addr: usize,
+    pub dma_addr: usize,
+    pub size: usize,
+    pub is_pinned: bool,
+    pub is_bounce_buffer: bool,
+    pub is_physical: bool,
+    pub is_prepinned: bool,
+}
+
+// Trait for the DMA interface
+pub trait DmaInterface {
```

Comment:
Do you envision that this would replace other uses of bounce buffering? (for example, copying from private memory into shared memory for isolated VMs, or when the block disk bounces for arm64 guests)?

Reply:
yes, i do envision that. 
Policy on GlobalDmaManager can be control the behavior system wide.

Reply:
Got it. How do you envision handling the case:

1. Have a VTL0 VM where sometimes memory needs to be pinned.
2. This particular transaction's memory must be placed in a bounce buffer, even if pinning would otherwise succeed?

(I'm thinking about the block device driver here, where it would never want to pin memory - the kernel doesn't know about the VTL0 addresses.)